Add primer on HTTP vs Gateway bots (#209)
IanMitchell/interaction-kit: README.md

Interaction Kit is a framework for creating Discord slash command bots over the HTTP API.

## HTTP vs Gateway

> **TL;DR**: you probably should be using [discord.js] unless you know why you're here.

Reviewer: Were you wanting to link this and forgot the link part of it?
`[discord.js](https://discord.js.org)`

Author: There's a bibliography entry at the end of the file that makes `[discord.js]` sufficient (as can be seen [here][1]), but I'll make it an inline link if Ian would prefer that.

[1]: https://github.com/stalruth/interaction-kit/tree/http-primer

Reviewer 2: I'd vote for inline since people are more used to it

> There are two ways for bots to recieve events from Discord.  Most API wrappers such as [discord.js] use a WebSocket connection (the "Gateway") to receive events, but Interaction Kit receieves Interaction events over HTTP.  As such, there are some major points to keep in mind before deciding to use Interaction Kit.

Reviewer: ```diff
- receieves
+ receives
```

### Gateway

The Gateway is a WebSocket that a bot maintains with Discord whenever it is online, giving it events as they happen.  The bot is able to subscribe to a wide array of events through Gateway Intents, allowing it to be notified on events such as message sends, a member changing their nickname or a guild changing its icon.

This is the only way to receive the majority of events Discord provides, and when dealing with users or guilds you can generally be sure you are operating on current information with sufficient caching.

The main drawback of the Gateway is that it requires your bot to maintain a connection with Discord at all times.  While there are affordable hosts for smaller scale bots, scaling can become expensive quickly and some developers have very low budgets.

### HTTP

> HTTP-only bots are sent Interaction events through HTTP requests sent from Discord.  These generally have a higher latency than a WebSocket connection from the same host and cannot listen for other types of event.

Reviewer: Perhaps consider putting Interaction in quotes to set it apart as a new term rather than what might seem like a fancy name for an event for Interaction Kit?

Author: I think a better option would be
```diff
- WebSocket connection (the "Gateway")
+ WebSocket connection called a Gateway
```
As that would make all references to Discord concepts consistent with each other.

> As a result of this HTTP-only bots lack real time updates to members and guilds, so unless a bot utilises REST API requests to get the latest data the chance that any cached data will be stale is significant.

Reviewer: I would add a comma after "As a result of this"

> As a result of this, HTTP-only bots lack real time updates to members and guilds, so unless a bot utilises REST API requests to get the latest data the chance that any cached data will be stale is significant.

Reviewer 2: ```suggestion
As a result of this, HTTP-only bots lack real time updates to members and guilds, so unless a bot utilizes REST API requests to get the latest data the chance that any cached data will be stale is significant.
```

> On the other hand, they are much more scalable than Gateway-based bots, with a variety of "serverless" plans on the market allowing HTTP-only bots to scale more easily and affordably than Gateway bots, with generous free plans that allow smaller bots to get off the ground without major infrastructure costs. HTTP-only bots without a bot user are also not subject to the 100 server pre-verification limit, however these bots are not able to be Verified at this time.

Reviewer 2: ```suggestion
On the other hand, they are much more scalable than Gateway-based bots, with a variety of "serverless" or "edge" plans on the market allowing HTTP-only bots to scale more easily and affordably than Gateway bots, with generous free plans that allow smaller bots to get off the ground without major infrastructure costs. 
```

The user bot part is accurate, but I'm gonna try to avoid mentioning that bit and push ppl to always add a user bot to ikit apps

> It needs to be emphasised that this approach does not work for all cases, as most real-time data can only be accessed through the Gateway - before making the decision to make a bot HTTP-only, care needs to be taken that Gateway features are not required.

Reviewer: ```diff
- care needs to be taken that Gateway features are not required.
+ care needs to be taken to ensure that Gateway features are not required.
```

> It needs to be emphasised that this approach does not work for all cases, as most real-time data can only be accessed through the Gateway - before making the decision to make a bot HTTP-only, care needs to be taken to ensure that Gateway features are not required.

Reviewer 2: ```suggestion
It needs to be emphasized that this approach does not work for all cases, as most real-time data can only be accessed through the Gateway - before making the decision to make a bot HTTP-only, care needs to be taken to ensure that Gateway features are not required.
```

are you Australian or something
